Updating a slice or a map from the in-memory backing store results in a panic

While working with the Microsoft Graph SDK, I've noticed updating a slice or a map from the in-memory backing store results in a panic.

The issue should be easily reproducible with the following test case:
```go
func TestReplaceSlice(t *testing.T) {
	memoryStore := NewInMemoryBackingStore()
	err := memoryStore.Set("key", []string{"a", "b"})
	assert.Nil(t, err)
	err = memoryStore.Set("key", []string{"b", "c"})
	assert.Nil(t, err)
}
```

```
=== RUN   TestReplaceSlice
--- FAIL: TestReplaceSlice (0.00s)
panic: runtime error: comparing uncomparable type []string [recovered]
	panic: runtime error: comparing uncomparable type []string

goroutine 17 [running]:
testing.tRunner.func1.2({0x5e3360, 0xc000184d30})
	C:/Program Files/Go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	C:/Program Files/Go/src/testing/testing.go:1399 +0x39f
panic({0x5e3360, 0xc000184d30})
	C:/Program Files/Go/src/runtime/panic.go:884 +0x212
github.com/microsoft/kiota-abstractions-go/store.(*InMemoryBackingStore).Set(0xc00008dc38, {0x615457?, 0x2?}, {0x5d58a0?, 0xc00018a258})
	D:/git/kiota-abstractions-go/store/inmemory_backing_store.go:59 +0xc6
github.com/microsoft/kiota-abstractions-go/store.TestReplaceSlice(0x0?)
	D:/git/kiota-abstractions-go/store/inmemory_backing_store_test.go:14 +0x26b
testing.tRunner(0xc0001a6ea0, 0x640820)
	C:/Program Files/Go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	C:/Program Files/Go/src/testing/testing.go:1493 +0x35f
```

The simplest fix I can think of is replacing the direct comparison with a call to `reflect.DeepEqual`, but performance-wise it might be better to implement a custom-written function.

--- a/store/inmemory_backing_store.go
+++ b/store/inmemory_backing_store.go
@@ -3,6 +3,7 @@ package store
 import (
 	"errors"
 	"github.com/google/uuid"
+	"reflect"
 	"strings"
 )
 
@@ -56,7 +57,7 @@ func (i *InMemoryBackingStore) Set(key string, value interface{}) error {
 	current := i.store[key]
 
 	// check if objects values have changed
-	if current == nil || current != value {
+	if current == nil || hasChanged(current, value) {
 		// track changed key
 		i.changedValues[key] = i.GetInitializationCompleted()
 
@@ -70,6 +71,15 @@ func (i *InMemoryBackingStore) Set(key string, value interface{}) error {
 	}
 
 	return nil
+}
+
+func hasChanged(current interface{}, value interface{}) bool {
+	kind := reflect.ValueOf(current).Kind()
+	if kind == reflect.Map || kind == reflect.Slice || kind == reflect.Struct {
+		return !reflect.DeepEqual(current, value)
+	} else {
+		return current != value
+	}
 }
 
 func (i *InMemoryBackingStore) Enumerate() map[string]interface{} {

--- a/store/inmemory_backing_store_test.go
+++ b/store/inmemory_backing_store_test.go
@@ -4,6 +4,7 @@ import (
 	"github.com/google/uuid"
 	"github.com/microsoft/kiota-abstractions-go/serialization"
 	"github.com/stretchr/testify/assert"
+	"reflect"
 	"testing"
 )
 
@@ -149,6 +150,58 @@ func TestClear(t *testing.T) {
 	testUser.GetBackingStore().Clear()
 
 	assert.Nil(t, testUser.GetId())
+}
+
+func TestReplaceSlice(t *testing.T) {
+	memoryStore := NewInMemoryBackingStore()
+	assert.Equal(t, 0, len(memoryStore.Enumerate()))
+
+	err := memoryStore.Set("key", []string{"a", "b"})
+	assert.Nil(t, err)
+
+	err = memoryStore.Set("key", []string{"b", "c"})
+	assert.Nil(t, err)
+
+	assert.Equal(t, 1, len(memoryStore.Enumerate()))
+	val := memoryStore.Enumerate()["key"].([]string)
+	assert.Equal(t, 2, len(val))
+	assert.Equal(t, "b", val[0])
+	assert.Equal(t, "c", val[1])
+}
+
+func TestReplaceMap(t *testing.T) {
+	memoryStore := NewInMemoryBackingStore()
+	assert.Equal(t, 0, len(memoryStore.Enumerate()))
+
+	err := memoryStore.Set("key", map[string]string{"k": "v1"})
+	assert.Nil(t, err)
+
+	err = memoryStore.Set("key", map[string]string{"k": "v2"})
+	assert.Nil(t, err)
+
+	assert.Equal(t, 1, len(memoryStore.Enumerate()))
+	val := memoryStore.Enumerate()["key"].(map[string]string)
+	assert.Equal(t, 1, len(val))
+	assert.Equal(t, "v2", val["k"])
+}
+
+func TestReplaceStruct(t *testing.T) {
+	memoryStore := NewInMemoryBackingStore()
+	assert.Equal(t, 0, len(memoryStore.Enumerate()))
+
+	prev := struct{ slice []string }{
+		slice: []string{"a", "b"},
+	}
+	err := memoryStore.Set("key", prev)
+	assert.Nil(t, err)
+
+	curr := struct{ slice []string }{
+		slice: []string{"b", "c"},
+	}
+	err = memoryStore.Set("key", curr)
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(memoryStore.Enumerate()))
+	assert.True(t, reflect.DeepEqual(curr, memoryStore.Enumerate()["key"]))
 }
 
 type testEntity struct {

